tools/restatectl/src/connection.rs

```diff
@@ -0,0 +1,343 @@
+        default_value = "http://localhost:5122/",
+        env = "RESTATE_ADDRESS",
+        global = true
+    )]
+    pub addresses: Vec<AdvertisedAddress>,
+
+    /// Sync metadata from metadata store first
+    #[arg(long)]
+    pub sync_metadata: bool,
+
+    #[clap(skip)]
+    nodes_configuration: Arc<Mutex<Option<NodesConfiguration>>>,
+
+    #[clap(skip)]
+    logs: Arc<Mutex<Option<Logs>>>,
+
+    #[clap(skip)]
+    cache: Arc<Mutex<HashMap<AdvertisedAddress, Channel>>>,
+}
+
+impl ConnectionInfo {
+    /// Gets NodesConfiguration object. This function tries all provided addresses and makes sure
+    /// nodes configuration is cached.
+    pub async fn get_nodes_configuration(&self) -> Result<NodesConfiguration, ConnectionInfoError> {
+        if self.addresses.is_empty() {
+            return Err(ConnectionInfoError::NoAvailableNodes(NoRoleError(None)));
+        }
+
+        let guard = self.nodes_configuration.lock().await;
+
+        // get nodes configuration will always use the addresses seed
+        // provided via the cmdline
+        self.get_latest(
+            self.addresses.iter(),
+            self.addresses.len(),
+            MetadataKind::NodesConfiguration,
+            guard,
+            |ident| Version::from(ident.nodes_config_version),
+        )
+        .await
+    }
+
+    /// Gets Logs object.
+    ///
+    /// This function will try multiple nodes learned from nodes_configuration
+    /// to get the best guess of the latest logs version is.
+    pub async fn get_logs(&self) -> Result<Logs, ConnectionInfoError> {
+        let nodes_config = self.get_nodes_configuration().await?;
+
+        let guard = self.logs.lock().await;
+
+        let mut nodes_addresses = nodes_config
+            .iter()
+            .map(|(_, node)| &node.address)
+            .collect::<Vec<_>>();
+
+        nodes_addresses.shuffle(&mut thread_rng());
+
+        let cluster_size = nodes_addresses.len();
+        let cached = self.cache.lock().await.keys().cloned().collect::<Vec<_>>();
+
+        assert!(!cached.is_empty(), "must have cached connections");
+
+        let try_nodes = cached
+            .iter()
+            .chain(
+                nodes_addresses
+                    .into_iter()
+                    .filter(|address| !cached.contains(address)),
+            )
+            .collect::<Vec<_>>();
+
+        // To be sure we landed on the best guess of the latest version of the logs
+        // we need to ask multiple nodes in the nodes config.
+        // We make sure cached nodes has higher precedence + 50% of node set size trimmed to
+        // a total of 50% + 1 nodes.
+
+        self.get_latest(
+            try_nodes.into_iter(),
+            (cluster_size / 2) + 1,
+            MetadataKind::Logs,
+            guard,
+            |ident| Version::from(ident.logs_version),
+        )
+        .await
+    }
+
+    /// Gets NodesConfiguration object. This function will try `try_best_of` nodes from the provided
```

Comment:
Might be outdated that it only gets the `NodesConfiguration` since we seem to also fetch other metadata.